src/plugins/cordova-plugin-device-orientation/compass-widget.js

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+
+/**
+ * @param {object=} options
+ * @constructor
+ */
+function CompassWidget(options) {
+    options = options || {};
+
+    this._container = options.container || document.body;
+
+    this._canvasElement = document.createElement('canvas');
+    this._indicatorCanvasElement = document.createElement('canvas');
+
+    this._container.appendChild(this._indicatorCanvasElement);
+    this._container.appendChild(this._canvasElement);
+
+    this._context = this._canvasElement.getContext('2d');
+    this._diameter = (typeof options.diameter === 'number') ? options.diameter : CompassWidget.Defaults.DIAMETER;
+
+    this._wrapperSize = 25;
+    this._compassBorderSize = 15;
+
+    this._canvasElement.style.position = 'absolute';
+    this._canvasElement.style.cursor = 'pointer';
+    this._canvasElement.style.top = '12px';
+    this._canvasElement.style.left = '12px';
+    this._canvasElement.width = this._diameter;
+    this._canvasElement.height = this._diameter;
+
+    this._center = {
+        x: this._diameter / 2,
+        y: this._diameter / 2
+    };
+
+    this._pointer = {
+        width: 30,
+        height: this._diameter - this._compassBorderSize * 2,
+        offset: this._compassBorderSize
+    };
+
+    this._onDragStartCallback = this._onDragStart.bind(this);
+    this._onDraggingCallback = this._onDragging.bind(this);
+    this._onDragEndCallback = this._onDragEnd.bind(this);
+    this._onHeadingUpdatedCallback = null;
+
+    this._canvasElement.addEventListener('mousedown', this._onDragStartCallback);
+
+    this._canvasElement.addEventListener('click', function (event) {
+        this._updateHeadingToPosition(event.clientX, event.clientY);
+    }.bind(this));
+
+    this._heading = {
+        value: 0,
+        direction: CompassWidget.getDirection(this.value)
+    };
+
+    if (typeof options.headingUpdatedCallback === 'function') {
+        this._onHeadingUpdatedCallback = options.headingUpdatedCallback;
+    }
+}
+
+CompassWidget.Defaults = {
+    DIAMETER: 160
+};
+
+CompassWidget.Direction = {
+    N: 'N',
+    NE: 'NE',
+    E: 'E',
+    SE: 'SE',
+    S: 'S',
+    SW: 'SW',
+    W: 'W',
+    NW: 'NW'
+};
+
+/**
+ * Get the direction according to the heading value.
+ * @param {number} heading A number from 0 to 359.99.
+ * @return {string} direction It can be one of the following: N, NE, E, SE, S, SW, W or NW.
+ */
+CompassWidget.getDirection = function (heading) {
```

Comment:
I don't like to have this function duplicated in both plugins: device-orientation and geolocation.
Both plugins needs to be synchronized about the heading value (orientation), so when these ranges needs to be updated, we need to change it for both simulation plugins.
What do you think about adding this function to a common module, like an utility, that is shared between all plugins and provided by cordova-simulate? By provided by cordova-simulate I mean that, in a future it could be really good to see this simulation logic implemented as part of the plugin itself and in the plugin repository, but the utilities modules provided by cordova simulate.
If we follow that approach, we share the same ranges to determined the heading value.